### PR TITLE
PLT-3636 Modifying permissions system to use roles which grant permissions.

### PR DESCRIPTION
**Pull request by crspeller**
#### Summary
Part one of guest accounts. This PR is for the required changes to the permissions system. 

Enterprise PR: https://github.com/mattermost/enterprise/pull/52

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has enterprise changes (please link)
- [x] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by enahum** [comment]
+1 this is great

File: model/client.go [hunk 1]
```diff
@@ -975,6 +975,7 @@ func (c *Client) CreateDirectChannel(userId string) (*Result, *AppError) {
 	if r, err := c.DoApiPost(c.GetTeamRoute()+"/channels/create_direct", MapToJson(data)); err != nil {
 		return nil, err
 	} else {
+		l4g.Debug(r.Body)
```
Comment (jwilander) [inline model/client.go]: Missed removing this?

**Comment by jwilander** [comment]
+1 other than minor comment

File: model/client.go (hunk 1 above)
Comment (crspeller) [inline model/client.go]: Good catch!

**Comment by jwilander** [comment]
I think there's a rebase issue

**Comment by jwilander** [comment]
re +1

**Comment by enahum** [comment]
re +1

**Pull request merged**